### PR TITLE
fix(middlewares): considers ports when checking if a host is excluded from corporate proxy

### LINKED ISSUE #722: BUG - checking if a host is excluded from corporate proxy not working properly
As described in the [answers.sap.com](https://answers.sap.com/questions/13728854/using-fiori-tools-proxy-behind-a-corporate-proxy-n.html?source=email-global-notification-qa-new-in-tag-followed) the check if a host is excluded from the corporate proxy fails if the host has  a port, e.g. `https://www.sap.com:443`, but the entry in the `no_proxy` list is e.g. `*.sap.com`. This is not correct as per the `no_proxy` definition:
```
no_proxy is a comma- or space-separated list of machine
or domain names, with optional :port part.  If no :port
part is present, it applies to all ports on that domain.
```

### Expected results
The check if a host is excluded from the corporate proxy works as expected

### Actual results
The check if a host is excluded from the corporate proxy doesn't works as expected

### Screenshots
If applicable, add screenshots to help explain the problem.

#### Version/Components/Environment	
Add any other context about the problem here
OS:                      
- [x] Mac OS
- [x] Windows
- [ ] Other

### Problem
{describe the problem} 

### Fix
{describe the fix} 

### Why was it missed
{Some explanation why this issue might have been missed during normal development/testing cycle}  

### How can we avoid this
{if we don’t want to see this type of issues anymore what we should do to prevent}

### PR DESCRIPTION
**Pull request by zdravko-georgiev**
Fixes #722

### CODE REVIEW

**Comment by zdravko-georgiev** [comment]
Question in the round: shall we create a `proxy-utils` package to get rid of the code duplication?

**Comment by tobiasqueck** [comment]
> Question in the round: should we create a `proxy-utils` package to get rid of the code duplication?

Can you give me a little time with #687. It might make sense to create a `ui5-extension-base` or so collecting reusable code for tasks and middlewares. 
Btw, I am strongly against a `-utils` or `-helper` module. They quickly become a dump for everything (btw `-base` is also not much better)

**Comment by tobiasqueck** [comment]
Summary of a discussion with @zdravko-georgiev :
* the proxy code gets more and more complicated and also still has one known issue
* instead of maintaining it ourselves, lets use an existing library: https://www.npmjs.com/package/proxy-from-env - it has a small footprint and a large number of downports
* its exported functionality does not match exactly what we need but one of the private functions
* @zdravko-georgiev will open a PR and suggest to expose it as well
* in the meantime we could use the fork of it here

**Comment by zdravko-georgiev** [comment]
@tobiasqueck Thanks for the summary!

File: packages/ui5-proxy-middleware/src/base/proxyFromEnv.d.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import * as _proxyFromEnv from 'proxy-from-env';
```
[the remark this answers is not in the document]
Comment (zdravko-georgiev) [reply packages/ui5-proxy-middleware/src/base/proxyFromEnv.d.ts:1]: Please see above

[the remark this answers is not in the document]
Comment (tobiasqueck) [reply packages/ui5-proxy-middleware/src/base/proxyFromEnv.d.ts:1]: The import isn't required. Your statement below adds the function but doesn't overwrite anything else.

File: packages/backend-proxy-middleware/src/base/proxyFromEnv.d.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import * as _proxyFromEnv from 'proxy-from-env';
```
[the remark this answers is not in the document]
Comment (Klaus-Keller) [reply packages/backend-proxy-middleware/src/base/proxyFromEnv.d.ts:1]: In this case, what about just doing
```typescript
import 'proxy-from-env';
```
?

File: packages/ui5-proxy-middleware/src/base/proxyFromEnv.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (zdravko-georgiev) [reply packages/ui5-proxy-middleware/src/base/proxyFromEnv.d.ts:1]: @tobiasqueck if I remove `import * as _proxyFromEnv from 'proxy-from-env';` and have only in the file

```
declare module 'proxy-from-env' {
    function shouldProxy(hostname: string, port: number): boolean;
}
```
Then when I try to do `import { getProxyUrl } from 'proxy-from-env'` I get the error message that `there is no exported member getProxyUrl`. So I am overriding everything else, right?

File: packages/backend-proxy-middleware/src/base/proxyFromEnv.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Klaus-Keller) [reply packages/backend-proxy-middleware/src/base/proxyFromEnv.d.ts:1]: also, if we keep the approach of forking and declaring the function, what about  renaming this file to `proxy-from-env.d.ts`?

[the remark this answers is not in the document]
Comment (zdravko-georgiev) [reply packages/backend-proxy-middleware/src/base/proxyFromEnv.d.ts:1]:
```TS
import 'proxy-from-env';
```
This one works 😃

[the remark this answers is not in the document]
Comment (zdravko-georgiev) [reply packages/backend-proxy-middleware/src/base/proxyFromEnv.d.ts:1]: > also, if we keep the approach of forking and declaring the function, what about renaming this file to `proxy-from-env.d.ts`?

👍

File: packages/ui5-proxy-middleware/src/base/proxyFromEnv.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (zdravko-georgiev) [reply packages/ui5-proxy-middleware/src/base/proxyFromEnv.d.ts]: Removed with https://github.com/SAP/open-ux-tools/pull/727/commits/11f244bcc480cda3b5ab40d71c09dfb6dc962cfa

File: packages/backend-proxy-middleware/src/base/proxyFromEnv.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (zdravko-georgiev) [reply packages/backend-proxy-middleware/src/base/proxyFromEnv.d.ts]: File removed with https://github.com/SAP/open-ux-tools/pull/727/commits/11f244bcc480cda3b5ab40d71c09dfb6dc962cfa

**Review by tobiasqueck — APPROVED** [review]
* code changes adhere to our coding guidelines
* updated test cases cover the modified functionality
* README updated to explain in more detail how the proxy works

File: packages/backend-proxy-middleware/src/base/config.ts [hunk 3]
```diff
@@ -1,44 +1,19 @@
...
-        process.env.FIORI_TOOLS_PROXY ||
-        process.env.http_proxy ||
-        process.env.HTTP_PROXY ||
-        process.env.https_proxy ||
-        process.env.HTTPS_PROXY ||
-        process.env.npm_config_proxy ||
-        process.env.npm_config_https_proxy
-    );
-}
+    const proxyFromFioriToolsConfig = proxyFromArgs || proxyFromConfig || process.env.FIORI_TOOLS_PROXY;
```
Comment (Klaus-Keller) [inline packages/backend-proxy-middleware/src/base/config.ts:13]: I wonder whether this causes that values set in `process.env` are overwritten by config from `ui5.yaml`.

**Review by Klaus-Keller — APPROVED** [review]
- I've rechecked that good and it looks good
- Test coverage is great
- Installed and built locally
- One open question about the priority of configurations, but I think this should be addressed separately

Approved, thank @zdravko-georgiev

**Comment by zdravko-georgiev** [comment]
Thanks @Klaus-Keller

**Pull request merged**